### PR TITLE
Added reduce= operator.

### PR DESCRIPTION
**Pull request by vasslitvinov**
This operator represenst the accumulation operation for a reduce intent.

* Added reduce= token.

* Parser converts "x reduce= y" to: PRIM_REDUCE_ASSIGN(x,y)

* At start of normalization, it is lowered to the call:
globalOp.accumulateOntoState(x,y)

* Added a check that there are no PRIM_REDUCE_ASSIGN calls in the tree
after resolution. Really, there shouldn't be any after normalization.
We do not already have a loop over all gCallExprs in the post-normalization
check, and I did not want to add one.

* Added accumulateOntoState() to our standard ReduceScanOp subclasses.

* While there, I removed the default value for the char* 'op' argument of
buildAssignment() - because that default was unused and, more importantly,
there is an assertion against that default in the function itself.

Tested: linux64; multilocale tests under gasnet.

### CODE REVIEW

**Comment by vasslitvinov** [comment]
@benharsh or @mppf would you review?

**Comment by mppf** [comment]
My first feedback, purely from the PR message, is why do we need a 'reduce=' token? I would have expected that we'd extend the parser to allow arbitrary functions like myoperation= . However I suppose you might need PRIM_REDUCE_ASSIGN anyway for this effort and the reduce= token might be easier?

File: compiler/include/bison-chapel.h [hunk 1]
```diff
@@ -1,19 +1,19 @@
-/* A Bison parser, made by GNU Bison 3.0.4.  */
+/* A Bison parser, made by GNU Bison 2.5.  */
```
Comment (mppf) [inline compiler/include/bison-chapel.h]: Next time you make a change that involves rebuilding the parser, put the parser rebuild in a separate commit on its own. That makes ignoring it easier when reviewing.

File: compiler/passes/normalize.cpp [hunk 2]
```diff
@@ -276,6 +278,82 @@ static void transformLogicalShortCircuit()
   }
 }
 
+
+//
+// lowerReduceAssign(): lower the reduce= calls, verifying their correct use.
+//
```
Comment (mppf) [inline compiler/passes/normalize.cpp]: I'd expect the above comment to be next to lowerReduceAssign. Also, I'd really like the comment to describe what reduce= starts out as and what it is "lowered" into.

File: compiler/passes/normalize.cpp [hunk 3]
```diff
@@ -276,6 +278,82 @@ static void transformLogicalShortCircuit()
...
+
+      // l.h.s. must be a single variable
+      if (SymExpr* lhsSE = toSymExpr(call->get(1))) {
+        Symbol* lhsVar = lhsSE->var;
+        // ... which is mentioned in a with clause with a reduce intent
+        Expr* curr = call->parentExpr;
+        BlockStmt* enclosingForall = NULL;
+        while (curr) {
+          if (BlockStmt* block = toBlockStmt(curr))
+            if (block->byrefVars) {
```
Comment (mppf) [inline compiler/passes/normalize.cpp]: I'm unfamilier with BlockStmt -> byrefVars but it looks like it is storing a PRIM_FORALL_LOOP in which case it really ought to be renamed. As it is I am confused why lowering reduce= needs to concern itself with any variable called "byrefVars".

File: modules/internal/ChapelReduce.chpl [hunk 4]
```diff
@@ -87,6 +87,9 @@ module ChapelReduce {
     proc accumulate(x) {
       value += x;
     }
+    proc accumulateOntoState(ref state, x) {
```
Comment (mppf) [inline modules/internal/ChapelReduce.chpl]: is this PR changing the interface for custom reductions?

File: test/parallel/forall/vass/reduce-assign-1.chpl [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  writeln("|| reduce:   ", lOr);
+
+}
+
+class PlusReduceOp: ReduceScanOp {
+  type eltType;
+  var  value: eltType;
+  proc identity         return 0: eltType;
+  proc accumulate(elm)  { value = value + elm; }
+  proc accumulateOntoState(ref state, elm) { state = state + elm; }
```
Comment (mppf) [inline test/parallel/forall/vass/reduce-assign-1.chpl]: I'm confused about the change to the user-defined reductions API. In particular, why do we have both accumulate and accumulateOntoState? While transitional work might be warranted, I'd like to have at least a plan for the future reductions API that is not redundant. Did the discussion about input, output, and state types and their relationships impact the reduction API as well?

**Comment by mppf** [comment]
@vasslitvinov - while I am happy with this as an intermediate step, I'm concerned that we're changing the reductions API without:
 * knowing what the complete future API should be and
 * updating user documentation / primers / etc

I'm assuming that accumulateOntoState is only necessary if reduce= is actually used; and yet I hope we are aiming towards getting reductions to be more consistent and implemented with fewer code paths - meaning that in the future we probably want accumulateOntoState (or whatever it turns into) to replace accumulate.

In some ways I am surprised by this change because I had always viewed the reduction class itself as the "state" but here you are introducing a separate state type. That is probably needed (based on our other discussions) but what I am missing is how the separate state type fits into a bigger design picture. Can you explain somewhere why .accumulate() isn't enough? Do you already have some kind of plan for what the reductions API will need to be in the future?

**Comment by vasslitvinov** [comment]
Superseded by #4565
